3Delight shader menu fix

gkocov opened the pull request:
Generally describe what this PR will do, and why it is needed

- Small fix to the 3Delight ShaderMenu.py file, so that certain shaders that were previously not listed are now available (for ex. the Area Light shader, the Decay Light Filter, the Gobo Light Filter, etc.) The newly included shaders are listed under the 3Delight / Shader / Other menu. 

### Checklist ###

- [x] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [x] I have updated the documentation, if applicable.
- [x] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [x] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
Thanks Goran! This makes some nice shaders available from the menu, along with some that I think may be better left out or relocated. There are also a few I'm interested in getting your 3Delight expertise to understand how they should be used.

I hope I'm not being annoyingly picky. My goal is to save user frustration by making sure we don't offer up shaders that won't work as expected, not to be a shader-blocking bridge troll. 🦖

Here are all of the shaders I see in the `Other` menu now and some comments / questions :

1. Anisotropic - nice addition, would it make sense going in `3Delight.Shader.Surface`?
2. Area light - very nice addition :) This allows using geometry as a mesh light with no additional changes since 3Delight just uses shaders for lights. Would it be better to have it in a new `Light` menu? There may be a couple more that could be added to that menu to make it not an only child.
3. Blinn - nice addition, would it make sense going in `3Delight.Shader.Surface`?
4. Directional Light - Is this covered already by `3Delight.Light.DistantLight`? They look like they have the same parameters, do you prefer one over the other, or like both? Whichever we go with, it will need to be a Gaffer light instead of a regular shader. It can be added after https://github.com/GafferHQ/gaffer/blob/6f240e6287ac6fd340a2b1395b377b8bdbb146ce/startup/gui/menus.py#L197 to make it available as a Gaffer light.
5. Displacement Shader - Is this better than the `dlDisplacement` shader from the `3Delight.Shader.Displacement` menu?
6. Distant Light - Available from the `3Delight.Light.DistantLight` menu item.
7. Decay Filter - This is an cool one, is connecting it to a light's color parameter the way it's meant to be used? Might be a good companion to the area light in a `3Delight.Shader.Light` menu?
8. Gobo Filter - Also cool! Same question as the Decay Filter about putting it in `3Delight.Shader.Light`.
9. Lens Distortion - This one might require a bit more to get it working. We don't handle 3Delight's `lensshader` parameter as far as I can tell, so it won't work just yet.
10. Sky - This would be great to have, but I haven't figured out how to get it to work. Does it need to be plugged into an environment light, or is it a self-contained light? I could get a background but no illumination if I add it as a self-contained light, like I noted above about `menus.py`. If it should be plugged into an environment, we may need to do some special case loading of the shader, or see if there needs to be a change in the 3Delight shader. The `.oso` file shows it is a `Surface`, so we only add a single `out` plug for its output. That means in Gaffer it can't be plugged into the color plug on the environment shader. How do you typically use this shader?
11. Environment Light - Available from the `3Delight.Light.EnvironmentLight` menu item.
12. Env Sphere - Does this give the same functionality as the Environment Light? Or is it a candidate for adding to the light types in `menus.py`?
13. Hlight_sky - Same questions as the Sky Light.
14. Lambert - Nice addition, would it make sense going in `3Delight.Shader.Surface`?
15. Matte Object - This could be cool too but we need to figure out how to connect it. The plugs it accepts as inputs we are loading as `ClosurePlug` but we load `Surface` shaders as having a single `Plug` output, not a `ClosurePlug`. @johnhaddon correct me if I'm wrong, but changing lots of shader outputs to `ClosurePlug` sounds like a recipe for angry users after their graphs fall apart? Maybe we could support this by another special case change of the matte object shader's inputs to regular `Plug`?
16. Point Light - Available from the `3Delight.Light.PointLight` menu item.
17. Projection - I don't think this will work currently, we probably need to do something like https://github.com/GafferHQ/gaffer/pull/5540 to plug the camera name in appropriately.
18. Spot Light - Available from the `3Delight.Light.SpotLight` menu item.
19. Surface Shader - This is a curious one, I don't seem to be able to get it to do anything. Do you know how to use it?

ericmehl commented:
> 17\. Projection - I don't think this will work currently, we probably need to do something like...

This shouldn't be that hard actually, I was thinking it would need a camera node / object, but it's just a string. We will need to take care of the matrix parameter, it currently generates a warning `WARNING [ParameterList] Attribute "placementMatrix" has unsupported datatype "M44fData".

Though after trying this shader out, I'm not getting anything in the viewport. I have to plug it directly into a `ShaderAssignment` - it's one of those where the output is an OSL surface instead of a shader, so I can't plug its output into a surface like Principled shader.

Do you know how to connect up the shaders for projection?

gkocov commented:
Thanks for the detailed response Eric and I completely agree with you that there is more work needed to get the best user experience with using 3Delight's shader system in Gaffer. Before proceeding to discuss the shader system itself, I wanted to give a bit of a background to my general approach and reasoning to the Gaffer / 3Delight integration work that I've been doing recently (both this PR and the other ones I submitted the past week or so). Since 3Delight's integration in Gaffer is in many ways in a beta/heavily WIP state, my main goal was to add features that although might not have the most elegant implementation or might not provide optimal user experience, they hopefully enable power users to get as much as possible out of 3Delight within Gaffer. This unfortunately means that until implemented in a better way some things will be somewhat unintuitive or will require a somewhat hacky approach to accomplish the desired result. 

Coming back to 3Delight's shading system in Gaffer, due to certain limitations in how some 3Delight shaders are implemented and also due to certain limitations in how OSL shaders work in Gaffer, there are quite a few things that don't work as expected out of the box: 

- In Gaffer currently OSL shaders with closure input parameters (for ex. 3Delight's Layered Material shader) don't accept connections from closure output parameters of OSL surface shaders, only from OSL generic shaders. Due to this it's not possible to connect any of the default surface shaders that come with 3Delight (dlPrincipled, dlGlass, dlMetal, dlStandard, etc.) to the closure inputs of dlLayeredMaterial or dlRandomMaterial. In order to overcome this limitation, a hacky but functional solution is to change the shader type in the compiled .oso files that come with 3Delight from surface to shader for dlPrincipled, dlGlass, dlMetal, dlStandard, etc. 
- The displacement shaders that come with 3Delight (dlDisplacement and dlDisplacementShader) are of generic shader type, not displacement type. When connected to a ShaderAssignment node in Gaffer they are registered as osl:surface shaders in the object's Gaffer attributes. In order to have them register as osl:displacement a similar hack can be done by changing the shader type in the compiled .oso files from shader to displacement. 
- Shaders with array input parameters that are declared without a set length don't seem to work as expected in Gaffer. Because of this the dlAOVGroup 3Delight shader which is used to define custom AOVs when connected to a surface shader's AOV Group input parameter can't be used. As a workaround I adapted the dlAOVGroup OSL shader source found here https://gitlab.com/3Delight/3delight-for-houdini/-/blob/master/osl/dlAOVGroup.osl to use a fixed number of inputs (four in my shader's case) instead of an array. 
 
The dlAOVGroupFour shader, the surface to shader converted dlPrincipled and shader to displacement converted dlDisplacement shaders can be found in my custom Gaffer build here https://drive.google.com/file/d/1BT68TPtv8dE9ykEswhy_gkSp93CEuXg0/view?usp=drive_link

For the moment I haven't looked into Gaffer's source in regards to how best to fix the above three issues. 

I apologize for the very long and very generic reply, but I felt that it might be good to go over that before moving to reply directly on the individual shaders that were exposed with my shaderMenu code change. 

1. Anisotropic - Completely agreed. Will look into moving it to 3Delight.Shader.Surface. 
2. Area Light - Completely agreed. Will look into creating a 3Delight.Shader.Light menu and moving it there. 
3. Blinn - Completely agreed. Will look into moving it to 3Delight.Shader.Surface. 
4. Directional Light - This is a duplicate of distant light (the compiled .oso code is exactly the same excluding the name). Can be removed from the 3Delight.Shader menu and no need to add it to the 3Delight.Light menu. 
5. Displacement Shader - Not really better than the Displacement shader that we already have. Slightly different and with less options. I can move it to 3Delight.Shader.Displacement, but as mentioned previously these shaders don't currently work correctly in Gaffer. I can remove it from the menu until we have a solution for the displacement shaders that come with 3Delight. 
6. Distant Light - Agreed that we can remove this one from the 3Delight.Shader menu since it's already present correctly as a light shader in 3Delight.Light. 
7. Decay Filter - Light filters are used by connecting the filter output to the filter input parameter of light shaders. Agreed that it would be good to move it to a new 3Delight.Shader.Light menu
8. Gobo Filter - Same
9. Lens Distortion - You are correct that currently we can't use lens shaders without large hacks like using an evaluate file or post processing an exported NSI scene. We can remove the shader from the menu until the time when there is support for lens shaders. 
10. Sky - This shader is declared as a surface shader and due to this doesn't show the color output parameters in Gaffer. Changing the .oso shader type from surface to shader gives the correct outputs, but I can remove it from the menu until we have a proper solution in Gaffer. 
11. Environment Light - Agreed that we can remove this one from the 3Delight.Shader menu since it's already present correctly as a light shader in 3Delight.Light. 
12. Env Sphere - I haven't used this shader in the past. I think it's not a light shader, but a spherical projection shader. 
13. Hlight_sky - This is a light shader (an environment light using a hemispherical sky texture as a source if I'm not mistaken). I'll look into adding it to menus.py, so that it's treated accordingly. 
14. Lambert - Completely agreed. Will look into moving it to 3Delight.Shader.Surface. 
15. Matte Object - This works in a similar way as dlLayerMaterial blending between two input closures using the Matte Opacity parameter. As mentioned previously the type of the shader used to connect to the Matte Object closure inputs needs to be generic shader, since surface is not supported currently in Gaffer. 
16. Point Light - Agreed that we can remove this one from the 3Delight.Shader menu since it's already present correctly as a light shader in 3Delight.Light. 
17. Projection - I haven't used it previously, so can't really say what would be best way to use it, sorry. 
18. Spot Light - Agreed that we can remove this one from the 3Delight.Shader menu since it's already present correctly as a light shader in 3Delight.Light. 
19. Surface Shader - I haven't used it in the past and similar to you I couldn't get it to output anything. I think we can remove it from the menu.

johnhaddon commented:
> In Gaffer currently OSL shaders with closure input parameters (for ex. 3Delight's Layered Material shader) don't accept connections from closure output parameters of OSL surface shaders, only from OSL generic shaders. Due to this it's not possible to connect any of the default surface shaders that come with 3Delight (dlPrincipled, dlGlass, dlMetal, dlStandard, etc.) to the closure inputs of dlLayeredMaterial or dlRandomMaterial.

I've opened a PR which I believe fixed the underlying problem here : https://github.com/GafferHQ/gaffer/pull/5629. If you're able to test that @gkocov, that would be great.

gkocov commented:
> I've opened a PR which I believe fixed the underlying problem here : #5629. If you're able to test that @gkocov, that would be great. 

Confirming that OSL surface shader outputs work correctly now and can be connected to closure inputs. Thanks so much John!

gkocov commented:
I apologize for the delay with the update, but I think I finally have a good solution to the above issues (and some additional ones that I found out while looking into how to solve the above ones). 

I have reworked slightly how we build the 3Delight/Shader sub menus. For the shaders that have tags, instead of assigning the shaders to predefined menus based on the tags, the code is now building sub menus inside 3Delight/Shader using the tag name. It's a subtle difference, but it simplifies the code slightly and more importantly it enables any user created custom shaders with custom tags to be assigned into their own respective menus and also provides forwards compatibility for any new shaders/tags that might be added to 3Delight in the future. 

I decided to keep the 'texture/2d' and 'texture/3d' tags to 'Texture 2D' and 'Texture 3D' for compatibility and to use the 'Light' menu for the 'Lightfilter' tag based on the above discussion. 

I've also added support for shaders that don't have tags. The tag-less shaders that come with 3Delight I'm either removing (if they don't work correctly or don't provide anything of importance) or adding to legacy menus based on the shader purpose (for ex. 'Texture 2D (Legacy) or Utility (Legacy)), with the exception of a few shaders that even though don't have tags provide functionality that is not available from the 3Delight shaders that do have tags (for ex. the Bump 2D shader)

The support for tag-less shaders also enables any custom user shaders that don't have tags to appear in the Other menu. 

There is some overlap between the new shaders (the ones with tags) and the legacy ones, but since it's difficult to say that the user won't run into a situation where it would be better to use the legacy ones, I decided to leave them in the menu especially since they already come with 3Delight (so the developers are OK with having them) and they are organized in Legacy menus, so the Gaffer user would know that the non-legacy shaders are usually a better choice. 

I've added the Hlight_sky shader (basically a combination of the Sky texture and the environmentLight shader) to the 3Delight lights in menu.py 

I've also added a couple of legacy shaders to the Displacement menu (Shading Engine_displacement and Builtin Displacement), both of which are of a displacement type so work correctly within Gaffer. The Shading Engine one enables the main Displacement shader to be connected to it and thus used correctly in Gaffer and Builtin Displacement is a simple displacement shader with much less options then the main Displacement shader, but it doesn't need the Shading Engine shader to work correctly. 

Finally I have added the dlAOVGroupFour shader source to the Gaffers OSL shaders inside a 3delight menu and tested that when compiled with Gaffer's OSL compiler it works correctly and provides custom AOVs when connected to a surface shaders AOV Group input plug.

ericmehl commented on shaders/3delight/dlAOVGroupFour.osl:
```diff
@@ -0,0 +1,51 @@
+	if(
+		raytype("camera") ||
+		(raytype("refraction") && visible_in_refractions) ||
+		(raytype("reflection") && visible_in_refractions) ||
```
I think this is meant to be `&& visible_in_reflections`.

gkocov replied:
Thanks for catching this! Fixed in the latest commit.

ericmehl commented on python/GafferDelightUI/ShaderMenu.py:
```diff
@@ -71,35 +71,158 @@ def __shaderSubMenu( oslPath ) :
+			elif shaderFileName == "dl3DelightMaterial" :
+				continue
+			elif shaderFileName == "transparent" :
+				continue
```
Small nitpick if you feel like changing it, it might be a bit more readable to do something like
```
if shaderFileName in [
    "distantLight",
    "directionalLight",
    "pointLight",
    ...
] :
    continue
```
instead of the repeated `elif`.

gkocov replied:
That code looks much better. Thanks for the suggestion and fixed in the latest commit.

ericmehl commented on python/GafferDelightUI/ShaderMenu.py:
```diff
@@ -71,35 +71,158 @@ def __shaderSubMenu( oslPath ) :
+			elif shaderFileName == "envChrome" :
+				subMenu = "Texture 3D (Legacy)"
+			elif shaderFileName == "snow" :
+				subMenu = "Texture 3D (Legacy)"
```
Similarly here, it might be more readable to do something like
```
subMenu = {
    "shadingEngine_displacement": "Displacement",
    "builtinDisplacement", "Displacement",
    ...
}.get( shaderFileName, None )
```

I probably should have done that myself in the earlier commit. But if you prefer the conditionals I won't hold up merging.

gkocov replied:
Thanks for this suggestion as well! Definitely looks better. I had to shift around some of the surrounding code since the above approach needs to be done before querying the tags and building the additional submenus from them, but I think the code ended up nicely organized at the end.

ericmehl commented on python/GafferDelightUI/ShaderMenu.py:
```diff
@@ -71,35 +71,158 @@ def __shaderSubMenu( oslPath ) :
+				continue
+			elif shaderFileName == "directionalLight" :
+				continue
+			elif shaderFileName == "distantLight" :
```
Duplicate entry for "distantLight" can be removed.

gkocov replied:
This is fixed now.

ericmehl commented on python/GafferDelightUI/ShaderMenu.py:
```diff
@@ -71,35 +71,158 @@ def __shaderSubMenu( oslPath ) :
+				subMenu = "Utility (Legacy)"
+			elif shaderFileName == "multiplyDivide" :
+				subMenu = "Utility (Legacy)"
+			elif shaderFileName == "Set Range" :
```
I think this should be "setRange"? It appears in "Utility" now, instead of "Utility (Legacy)".

gkocov replied:
Actually there are two set range shaders - dlSetRange.oso and setRange.oso. One operates on float values and the other one on vectors. I changed the Gaffer name of the float shader to Set Range Float, kept the other one as Set Range and placed both in the main Utility menu.

ericmehl reviewed (changes requested):
Thanks the reorganization Goran! Nice to have the older shaders marked as Legacy to make them available but tip off users that they may not be the latest-and-greatest.

I just made note of a couple of easy fixes and potential formatting improvements, but don't want to let the formatting prevent getting it merged.

gkocov commented:
Thanks for the notes! I committed a new version with all of the above items fixed, plus a couple of additional fixes. 

The old displacement shaders that come with 3Delight unfortunately don't work fully as expected (they displace the geo, but something is messed up with the recalculation of the normals after the displacement), so I changed strategy a bit. I noticed that we already have some code to adjust one shader type to another in src/GafferOSL/OSLShader.cpp (vdbVolume from shader to volume type). I piggy backed on that and added code for dlDisplacement, so that it is recognized as a displacement shader instead of a generic shader. This has worked great in my tests, so hopefully this solves the displacement shader issue finally. 

I removed the old displacement shaders from the menu along with the shading engine shaders, since they really didn't offer anything to the user.

gkocov commented:
Also as a minor note, I changed the name of the ShaderMenu.py shaderFileName variable to shaderNiceName so it's more accurate to what it actually holds.

ericmehl commented:
Fantastic, all looks good to me! Hitting merge...
